src/iop/colorbalancergb.c

```diff
@@ -668,6 +817,206 @@ static void paint_chroma_slider(GtkWidget *w, const float hue)
+  if(self->request_mask_display)
+  {
+    dt_control_log(_("cannot display masks when the blending mask is displayed"));
+    g->mask_display = 0;
+  }
+  else
+  {
+    g->mask_display = dt_bauhaus_widget_get_quad_active(GTK_WIDGET(togglebutton));
+  }
+
+  if(g->mask_display)
+  {
+    if(togglebutton == g->shadows_weight) g->mask_type = MASK_SHADOWS;
+    if(togglebutton == g->grey_fulcrum) g->mask_type = MASK_MIDTONES;
+    if(togglebutton == g->highlights_weight) g->mask_type = MASK_HIGHLIGHTS;
+  }
+  else
+  {
+    g->mask_type = MASK_NONE;
+  }
+
+  dt_bauhaus_widget_set_quad_active(GTK_WIDGET(g->shadows_weight), g->mask_type == MASK_SHADOWS);
+  dt_bauhaus_widget_set_quad_active(GTK_WIDGET(g->grey_fulcrum), g->mask_type == MASK_MIDTONES);
+  dt_bauhaus_widget_set_quad_active(GTK_WIDGET(g->highlights_weight), g->mask_type == MASK_HIGHLIGHTS);
+
+  dt_iop_refresh_center(self);
+}
+
+
+static gboolean dt_iop_tonecurve_draw(GtkWidget *widget, cairo_t *crf, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  dt_iop_colorbalancergb_params_t *p = (dt_iop_colorbalancergb_params_t *)self->params;
+  const float shadows_weight = 2.f + p->shadows_weight * 2.f;
+  const float highlights_weight = 2.f + p->highlights_weight * 2.f;
+
+  // Cache the graph objects to avoid recomputing all the view at each redraw
+  GtkAllocation allocation;
+  gtk_widget_get_allocation(widget, &allocation);
+  GtkStyleContext *context = gtk_widget_get_style_context(widget);
+
+  cairo_surface_t *cst =
+    dt_cairo_image_surface_create(CAIRO_FORMAT_ARGB32, allocation.width, allocation.height);
+  PangoFontDescription *desc =
+    pango_font_description_copy_static(darktable.bauhaus->pango_font_desc);
+  cairo_t *cr = cairo_create(cst);
+  PangoLayout *layout = pango_cairo_create_layout(cr);
+
+  const gint font_size = pango_font_description_get_size(desc);
+  pango_font_description_set_size(desc, 0.95 * font_size);
+  pango_layout_set_font_description(layout, desc);
+  pango_cairo_context_set_resolution(pango_layout_get_context(layout), darktable.gui->dpi);
+
+  char text[256];
+
+  // Get the text line height for spacing
+  PangoRectangle ink;
+  snprintf(text, sizeof(text), "X");
+  pango_layout_set_text(layout, text, -1);
+  pango_layout_get_pixel_extents(layout, &ink, NULL);
+  const float line_height = ink.height;
+
+  const float inset = DT_PIXEL_APPLY_DPI(4);
+  const float margin_top = inset;
+  const float margin_bottom = line_height + 2 * inset;
+  const float margin_left = line_height + inset;
+  const float margin_right = 0;
+
+  const float graph_width = allocation.width - margin_right - margin_left;   // align the right border on sliders
+  const float graph_height = allocation.height - margin_bottom - margin_top; // give room to nodes
+
+  gtk_render_background(context, cr, 0, 0, allocation.width, allocation.height);
+
+  // draw x gradient as axis legend
+  cairo_pattern_t *grad;
```

Comment:
could be all in one line:
cairo_pattern_t *grad = cairo_pattern_create_linear(margin_left, 0.0, graph_width, 0.0);

Reply:
Done